### ISSUE TITLE
How to determine whether there are currently running tasks

### ISSUE THREAD

**Issue by RichardSpace369** [redux-saga/redux-saga#1827]
When I execute component loadData on the server side, I need to make sure that loadData is executed before render
eg:
![image](https://user-images.githubusercontent.com/27414779/55851654-207b2080-5b8c-11e9-9c4f-830a7b3903d3.png)

**Comment by zhaoyingdu** [comment]
I wonder how you implemented your saga and how it is related to the loadData method. I think the answer to your issue is most likely depend on your design

here some side notes about your questions:   
1. "How to determine whether there are currently running tasks?"
A saga task is promisifiable, check this: [redux-saga/api#task](https://redux-saga.js.org/docs/api/#task)
2. "I need to make sure that loadData is executed before render"
If your loadData is not relevant to saga at all,
then if your loadData method is itself async, perhaps you don't have to enclose it in a promise. Or, if it is in callback style, you could promisify it first.

just some thoughts, might not be able to solve your problem.

**Comment by Pecking** [comment]
@zhaoyingdu 
Thanks, my loadData is asynchronous and related to redux-saga, I need to ensure that some asynchronous operations are performed before the render, but I don't know how to ensure that these asynchronous operations are completed, I tried using task.isRunning but He is always true, I don't know how to do it, I want to get help.

**Comment by zhaoyingdu** [comment]
**" I tried using task.isRunning but He is always true"**
if you are checking task on line 45, then, the problem is that, that task is referencing your root saga, as long as your root saga isn't returned, this task will always have task,isRunning() being true.

**"I need to ensure that some asynchronous operations are performed before the render"**
since loadData is async, then you should construct your promise list this way:  
In line 51, moving the call resolve() inside then, like this:
```javascript 
route.loadData(store).then(resolve()).catch(reject())
```
so your promise will resolve after loadData() is resolved.

**But**, you don't have to do this. because your loadData is already an async function. And I think this way of using promise is an antiPattern called "deffered promise", here is some explanations about [deffered anti pattern](https://medium.com/@7gravity/javascript-promise-anti-patterns-5dcdcf4bd2b6)
Basically, in this way, your Promise.all is resolving and catching the wrapper promises' state, but this complicates code, and has potential problem, like when you forgot to put .catch(reject()), your Promise.all will block forever

You could instead map it directly, like this:
```javascript
const asyncTask = mtRoutes.map(({route})=>{
  if(route.loadData){
    return route.loadData(store)
  }
)
```
and Promise.all will run and resolve all your route.loadData() promise functions, render(req,res,store) will be called after that

So I think your issue is likely related to javascript Promise for now.

**Comment by neurosnap** [comment]
Closing due to inactivity.  If this is still an issue then please reply and I will reopen and try to help.  Thanks!
